Adding more caching to GL backend.

icefoxen opened the pull request:
Like #1220 except based on the right starting point.  Sorry.

kvark commented on src/backend/gl/src/command.rs:
```diff
@@ -127,47 +134,188 @@ pub const RESET: [Command; 14] = [
+    }
+
+    fn bind_constant_buffer(&mut self, buf: &mut Vec<Command>, constant_buffer: c::pso::ConstantBufferParam<Resources>) {
+        if let Some(cb) = self.constant_buffer {
```
nit: could be simplified to `if self.constant_buffer == Some(constant_buffer)`
if it complains about the type being moved and such, you can compare by reference `if self.constant_buffer.as_ref() == Some(&constant_buffer)`

kvark commented on src/backend/gl/src/command.rs:
```diff
@@ -127,47 +134,188 @@ pub const RESET: [Command; 14] = [
     Command::SetBlendState(2, COLOR_DEFAULT),
     Command::SetBlendState(3, COLOR_DEFAULT),
-    Command::SetBlendColor([0f32; 4]),
+    Command::SetBlendColor([0f32; 4])
```
is that a `rustfmt` decision to remove the comma?

kvark commented on src/backend/gl/src/command.rs:
```diff
@@ -127,47 +134,188 @@ pub const RESET: [Command; 14] = [
     }
+
+    fn bind_resource_view(&mut self, buf: &mut Vec<Command>, resource_view: c::pso::ResourceViewParam<Resources>) {
+        if let Some(rv) = self.resource_view {
```
similarly, this can be rewritten

kvark commented on src/backend/gl/src/command.rs:
```diff
@@ -127,47 +134,188 @@ pub const RESET: [Command; 14] = [
+    }
+
+    fn bind_framebuffer(&mut self, buf: &mut Vec<Command>, access: Access, fb: FrameBuffer) {
+
```
extra line?

kvark commented on src/backend/gl/src/command.rs:
```diff
@@ -127,47 +134,188 @@ pub const RESET: [Command; 14] = [
+        // That's actually bad because it makes it impossible
+        // to completely remove all redundant calls if the
+        // stencil is enabled;
+        // we'll be re-enabling it over and over.
```
agreed

kvark commented on src/backend/gl/src/command.rs:
```diff
@@ -127,47 +134,188 @@ pub const RESET: [Command; 14] = [
+        buf.push(Command::SetStencilState(option_stencil, stencils, cullface));
+    }
+    fn set_blend_state(&mut self, buf: &mut Vec<Command>, color_slot: c::ColorSlot, color: s::Color) {
+        if let Some(bs) = self.blend_state {
```
nit: simplify

kvark commented on src/backend/gl/src/command.rs:
```diff
@@ -127,47 +134,188 @@ pub const RESET: [Command; 14] = [
+        }
+    }
+
+    fn push(&self, buf: &mut Vec<Command>, command: Command) {
```
what is the purpose of this method?

icefoxen replied:
Basically a layer of indirection.  It used to have a bit more functionality but that ended up getting pulled out into other functions.  Now it's basically there to say "the cache modifies the command buffer whenever necessary, nothing else needs to touch it directly."

kvark commented on src/backend/gl/src/command.rs:
```diff
@@ -127,47 +134,188 @@ pub const RESET: [Command; 14] = [
+        buf.push(command)
+    }
+
+    fn bind_program(&mut self, buf: &mut Vec<Command>, program: Program) {
```
I feel like this is a bit verbose on the `Cache` side. How about the `bind_*` methods returning `Option<Command>`, and the caller just doing something like `self.buf.extend(self.cache.bind_program(program))`? I wonder if we could chain those options to call `extend` for multiple results at once.

icefoxen replied:
A bit verbose, yes.  Like I've said before, this is a pretty brute-force attempt...

kvark commented on src/backend/gl/src/command.rs:
```diff
@@ -209,160 +358,199 @@ impl command::Buffer<Resources> for CommandBuffer {
-                (_, None) => (),
+        if self.cache.current_vbs == Some(vbs) {
+            return
+        } else {
```
no need for `else` branch here

kvark commented on src/backend/gl/src/command.rs:
```diff
@@ -209,160 +358,199 @@ impl command::Buffer<Resources> for CommandBuffer {
-                                          dst_offset_bytes as gl::types::GLintptr,
-                                          size_bytes as gl::types::GLsizeiptr));
+        self.cache.push(&mut self.buf,
+                                     Command::CopyBuffer(src,
```
indentation is broken here

kvark commented on src/backend/gl/src/command.rs:
```diff
@@ -209,160 +358,199 @@ impl command::Buffer<Resources> for CommandBuffer {
-                error!("GL: unable to update the contents of a Surface({})", s),
+            NewTexture::Texture(t) => {
+                self.cache.push(&mut self.buf,
+                                             Command::UpdateTexture(t, kind, face, ptr, img))
```
indentation here

kvark commented on src/backend/gl/src/command.rs:
```diff
@@ -371,22 +559,33 @@ impl command::Buffer<Resources> for CommandBuffer {
+                 count: c::VertexCount,
+                 instances: Option<command::InstanceParams>) {
+        self.cache.push(&mut self.buf,
+                                     Command::Draw(self.cache.primitive, start, count, instances));
```
indentation

kvark commented on src/backend/gl/src/command.rs:
```diff
@@ -371,22 +559,33 @@ impl command::Buffer<Resources> for CommandBuffer {
-            gl_index, RawOffset(offset as *const gl::types::GLvoid), count, base, instances));
+        self.cache
+            .push(&mut self.buf,
+                         Command::DrawIndexed(self.cache.primitive,
```
indentation

icefoxen commented:
What is the policy on rustfmt?  You don't appear to be using it as far as I can tell, so a lot of the indentation stuff is just emacs's auto-formatting.  If there's some specific formatting guide or such I should follow that'd make life easier.

kvark commented:
@icefoxen I haven't looked into `rustfmt` and there is no official policy. My rough understanding is that it's not ready for us yet, so no need to vigorously format anything :)

icefoxen commented:
All right.  I was more looking for ways to vigorously *not* reformat everything, since I usually code with rustfmt running automatically every time I save a file and it makes a terrible mess of projects that don't use it.

icefoxen commented:
Okay, I think I've fixed all the formatting bugs and simplified where indicated.  The only thing I haven't done is gotten rid of `Cache::push()` and altered the cache functions to return `Option` instead of maybe pushing the commands themselves.  I feel like the latter can work better because it gives the cache more control over what goes on, and makes the caller have to do less work.  All the caller has to do is call the `Cache`'s function to add a command.

I think that ideally every command should be added to the buffer via a method on `Cache`; if the cache can do nothing, it just pushes the command directly, but it would open the door to having more state be tracked by the `Cache` without having to change anything about how the caller operates.

kvark commented:
@icefoxen thanks for updating it!!!

> I feel like the latter can work better because it gives the cache more control over what goes on, and makes the caller have to do less work. All the caller has to do is call the Cache's function to add a command.
> I think that ideally every command should be added to the buffer via a method on Cache; if the cache can do nothing, it just pushes the command directly, but it would open the door to having more state be tracked by the Cache without having to change anything about how the caller operates.

Don't worry about the caller. There is only one, so it doesn't make much sense to try to shove more logic into the `Cache`, making it sort of an opaque layer that fills up the command buffer (AFAIK, that's what your PR is doing). Thing is, `CommandBuffer` IS this layer already, so what `Cache` needs to be is just a structure holding some state, possibly with a few helpers if that make sense.

Does this sound reasonable?

icefoxen commented:
Fair enough!  It's hard sometimes, you look at stuff and say "hey, a layer of abstraction could go here!" and why not put one in?

This will probably take me a few days to get around to doing, but I'll refactor it so the functions just return an option.

icefoxen commented:
Refactor complete; it IS nicer, thank you for the suggestion.

kvark commented:
Thanks @icefoxen ! Looks great now! 👍 
@homu r+

homu commented:
:pushpin: Commit 93febca has been approved by `kvark`

<!-- @homu r=kvark 93febca329486dd4f2204f68139449aef75a0967 -->

homu commented:
:broken_heart: Test failed - [status](https://travis-ci.org/gfx-rs/gfx/builds/223623209)

kvark commented on src/core/src/pso.rs:
```diff
@@ -208,7 +208,7 @@ impl Descriptor {
 
 /// A complete set of vertex buffers to be used for vertex import in PSO.
-#[derive(Copy, Clone, Debug)]
+#[derive(Copy, Clone, Debug, PartialEq)]
```
are these `core` changes required for your PR?
We need to bump the version numbers and (if core is changed) change the dependency of core to the new version.

icefoxen replied:
I'm afraid so because it needs to be able to compare the existing `VertexBufferSet` against the one it has cached.  It can probably be worked around by just storing the *contents* of the `VertexBufferSet`, but that seems a little redundant.

kvark replied:
Ok, I see. Please bump both version numbers then and make sure the new core is requested

kvark commented on src/backend/gl/src/command.rs:
```diff
@@ -141,33 +153,158 @@ pub const RESET: [Command; 14] = [
+    }
+
+    fn bind_program(&mut self, program: Program) -> Option<Command> {
+        if program == self.program {
```
nit: I think a simple `if/else` for the case of small branches should be preferred over `return`

kvark commented:
Ok, we can bump the versions ourselves and do the refactor, if needed.

kvark commented:
@icefoxen your code has sailed in https://crates.io/crates/gfx_device_gl/0.13.1
